Improve Lua API for Sound and Music

mpconley opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions

- Updating the Lua API for sound and music to manage media more effectively (details below).
- Adds "start" to Client.Media.Play parameters to begin sound and music play at the specified position in milliseconds.

#### Motivation for adding to Mudlet
Extending the capabilities of Mud Client Media Protocol (MCMP) to Mudlet's Lua API.
#### Other info (issues closed, discussion etc)
Recommend to close issues #2259, #2265, #2518, #2997, #3049, #3279 with merging this PR.
#### Release post highlight
<!--
Use this space if you wish to write a short statement or example for inclusion
in the release post for the next release. 
-->
Enhanced the Lua API for processing media files that extends all the capabilities Mud Client Media Protocol (MCMP) for script authors. Usage example:  Start background ambience music defined by a key in one script when in area/zone A and later when changing to a new area/zone replace the ambience music based on matching that same key.

Also added "start" to `Client.Media.Play` for GMCP to begin sound and music play at a specified position in the track. 

-Tamarindo@StickMUD

#### Details
Playing, Stopping and (Pre)Loading sounds and music have two flavors of API syntax:

1. **ordered parameters** - values for settings that must be listed in order
2. **table** - table format for setting key = value pairs (like volume = 75), listed in any order

**Playing Sound** - Play and track multiple sounds.
* ordered parameters - [playSoundFile(name[,volume][,fadein][,fadeout][,start][,loops][,key][,tag][,priority][,url])](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Play a sound file stored in the profile's media directory (i.e. /Users/Tamarindo/mudlet-data/profiles/StickMUD/media)
playSoundFile(
    "cow.wav" -- name
)
-- OR copy once from the game's profile, and play a sound file stored in the profile's media directory
---- volume of 75 (1 to 100)
playSoundFile(
    getMudletHomeDir().. "/cow.wav" -- name
    , 75 -- volume
)
-- OR copy once from the local file system, and play a sound file stored in the profile's media directory
---- volume of 75 (1 to 100)
playSoundFile(
    "C:/Users/Tamarindo/Documents/cow.wav" -- name
    , 75 -- volume
)
-- OR download once from the Internet, and play a sound stored in the profile's media directory
---- volume of 75
---- loops of 2 (-1 for indefinite repeats, 1+ for finite repeats)
---- key reference of "cow" for stopping this unique sound later
---- tag reference of "animals" to stop a group of sounds later with the same tag
---- priority of 25 (1 to 100, 50 default, a sound with a higher priority would stop this one)
---- url to download once from the Internet if the sound does not exist in the profile's media directory
playSoundFile(
    "cow.wav" -- name
    , 75 -- volume
    , nil -- fadein
    , nil -- fadeout
    , nil -- start
    , 2 -- loops
    , "cow" -- key 
    , "animals" -- tag
    , 25 -- priority
    , "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/" -- url
)
```
* table - [playSoundFile({parameter=value[,parameter=value]})](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Play a sound file stored in the profile's media directory (i.e. /Users/Tamarindo/mudlet-data/profiles/StickMUD/media)
playSoundFile({
    name = "cow.wav"
})
-- OR copy once from the game's profile, and play a sound file stored in the profile's media directory
---- volume of 75 (1 to 100)
playSoundFile({
    name = getMudletHomeDir().. "/cow.wav"
    , volume = 75
})
-- OR copy once from the local file system, and play a sound file stored in the profile's media directory
---- volume of 75 (1 to 100)
playSoundFile({
    name = "C:/Users/Tamarindo/Documents/cow.wav"
    , volume = 75
})
-- OR download once from the Internet, and play a sound stored in the profile's media directory
---- volume of 75
---- loops of 2 (-1 for indefinite repeats, 1+ for finite repeats)
---- key reference of "cow" for stopping this unique sound later
---- tag reference of "animals" to stop a group of sounds later with the same tag
---- priority of 25 (1 to 100, 50 default, a sound with a higher priority would stop this one)
---- url to download once from the Internet if the sound does not exist in the profile's media directory
playSoundFile({
    name = "cow.wav"
    , volume = 75
    , fadein = nil -- nil lines are optional, no need to use
    , fadeout = nil -- nil lines are optional, no need to use
    , start = nil -- nil lines are optional, no need to use
    , loops = 2
    , key = "cow"
    , tag = "animals"
    , priority = 25
    , url = "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/"
})
```

**Playing Music** - Play and track background music.
* ordered parameters - [playMusicFile(name[,volume][,fadein][,fadeout][,loops][,key][,tag][,continue][,url])](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Play a music file stored in the profile's media directory (i.e. /Users/Tamarindo/mudlet-data/profiles/StickMUD/media)
playMusicFile(
    "167124__patricia-mcmillen__rugby-club-in-spain.mp3"  -- name
)
-- OR copy once from the game's profile, and play a music file stored in the profile's media directory
---- volume of 75 (1 to 100)
playMusicFile(
    getMudletHomeDir().. "/167124__patricia-mcmillen__rugby-club-in-spain.mp3" -- name
    , 75 -- volume
)
-- OR copy once from the local file system, and play a music file stored in the profile's media directory
---- volume of 75 (1 to 100)
playMusicFile(
    "C:/Users/Tamarindo/Documents/167124__patricia-mcmillen__rugby-club-in-spain.mp3" -- name
    , 75 -- volume
)
-- OR download once from the Internet, and play music stored in the profile's media directory
---- fadein and increase the volume from 1 at the start position to default volume up until the position of 20 seconds
---- fadeout and decrease the volume from default volume to one, 53 seconds from the end of the music
---- start 10 seconds after position 0 (fadein scales its volume increase over a shorter duration, too)
---- key reference of "rugby" for stopping this unique music later
---- tag reference of "ambience" to stop and music later with the same tag
---- continue playing this music if another request for the same music comes in (false restarts it) 
---- url to download once from the Internet if the music does not exist in the profile's media directory
playMusicFile(
    "167124__patricia-mcmillen__rugby-club-in-spain.mp3" -- name
    , nil -- volume
    , 20000 -- fadein
    , 53000 -- fadeout
    , 10000 -- start
    , nil -- loops
    , "rugby" -- key 
    , "ambience" -- tag
    , true -- continue
    , "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/" -- url
)
```
* table - [playMusicFile({parameter=value[,parameter=value]})](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Play a music file stored in the profile's media directory (i.e. /Users/Tamarindo/mudlet-data/profiles/StickMUD/media)
playMusicFile({
    name = "167124__patricia-mcmillen__rugby-club-in-spain.mp3"
})
-- OR copy once from the game's profile, and play a music file stored in the profile's media directory
---- volume of 75 (1 to 100)
playMusicFile({
    name = getMudletHomeDir().. "/167124__patricia-mcmillen__rugby-club-in-spain.mp3"
    , volume = 75
})
-- OR copy once from the local file system, and play a music file stored in the profile's media directory
---- volume of 75 (1 to 100)
playMusicFile({
    name = "C:/Users/Tamarindo/Documents/167124__patricia-mcmillen__rugby-club-in-spain.mp3"
    , volume = 75
})
-- OR download once from the Internet, and play music stored in the profile's media directory
---- fadein and increase the volume from 1 at the start position to default volume up until the position of 20 seconds
---- fadeout and decrease the volume from default volume to one, 53 seconds from the end of the music
---- start 10 seconds after position 0 (fadein scales its volume increase over a shorter duration, too)
---- key reference of "rugby" for stopping this unique music later
---- tag reference of "ambience" to stop and music later with the same tag
---- continue playing this music if another request for the same music comes in (false restarts it) 
---- url to download once from the Internet if the music does not exist in the profile's media directory
playMusicFile({
    name = "167124__patricia-mcmillen__rugby-club-in-spain.mp3"
    , volume = nil -- nil lines are optional, no need to use
    , fadein = 20000
    , fadein = 53000
    , start = 10000
    , loops = nil -- nil lines are optional, no need to use
    , key = "rugby"
    , tag = "ambience"
    , continue = true
    , url = "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/"
})
```

**Stopping Sound** - Stop all sounds (no filter), or sound that meets a combination of filters (name, key, tag, and priority).
* ordered parameters (or none) - [stopSounds([name][,key][,tag][,priority])](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Stop all playing sound files for this profile associated with the API
stopSounds()
-- Stop playing the cow sound
stopSounds("cow.wav")
-- Stop playing any sounds tagged as "animals" with a priority less than or equal to 50
---- This would not stop sounds tagged as "animals" greater than priority 50.  This is an "AND" and not an "OR".
stopSounds(
    nil -- name
    , nil -- key
    , "animals" -- tag
    , 50 -- priority
)
```
* table - [stopSounds({parameter=value[,parameter=value]})](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Stop playing the cow sound
stopSounds({name = "cow.wav"})
-- Stop playing any sounds tagged as "animals" with a priority less than or equal to 50
---- This would not stop sounds tagged as "animals" greater than priority 50.  This is an "AND" and not an "OR".
stopSounds({
    name = nil -- nil lines are optional, no need to use
    , key = nil -- nil lines are optional, no need to use
    , tag = "animals"
    , priority = 50
})
```
**Stopping Music** - Stop all music (no filter), or music that meets a combination of filters (name, key, and tag).
* ordered parameters (or none) - [stopMusic([name][,key][,tag])](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Stop all playing music files for this profile associated with the API
stopMusic()
-- Stop playing the rugby mp3 by name
stopMusic("167124__patricia-mcmillen__rugby-club-in-spain.mp3")
-- Stop playing the unique sound identified as "rugby"
stopMusic(
    nil -- name
    , "rugby" -- key
    , nil -- tag
)
```
* table - [stopMusic({parameter=value[,parameter=value]})](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Stop playing the rugby mp3 by name
stopMusic({name = "167124__patricia-mcmillen__rugby-club-in-spain.mp3"})
-- Stop playing the unique sound identified as "rugby"
stopMusic({
    name = nil  -- nil lines are optional, no need to use
    , key = "rugby" -- key
    , tag = nil  -- nil lines are optional, no need to use
})
```

**Loading Sound** - Nice-to-have for preloading sounds.
* ordered parameters - [loadSoundFile(name[,url])](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Download from the Internet
loadSoundFile("cow.wav", "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/")
-- OR download from the profile
loadSoundFile(getMudletHomeDir().. "/cow.wav")
-- OR download from the local file system
loadSoundFile("C:/Users/Tamarindo/Documents/cow.wav")
```
* table - [loadSoundFile({parameter=value[,parameter=value]})](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Download from the Internet
loadSoundFile({
    name = "cow.wav"
    , url = "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/"
})
-- OR download from the profile
loadSoundFile({name = getMudletHomeDir().. "/cow.wav"})
-- OR download from the local file system
loadSoundFile({name = "C:/Users/Tamarindo/Documents/cow.wav"})
```

**Loading Music** - Nice-to-have for preloading music.
* ordered parameters - [loadMusicFile(name[,url])](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Download from the Internet
loadMusicFile(
    "167124__patricia-mcmillen__rugby-club-in-spain.mp3"
    , "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/"
)
-- OR download from the profile
loadMusicFile(getMudletHomeDir().. "/167124__patricia-mcmillen__rugby-club-in-spain.mp3")
-- OR download from the local file system
loadMusicFile("C:/Users/Tamarindo/Documents/167124__patricia-mcmillen__rugby-club-in-spain.mp3")
```
* table - [loadMusicFile({parameter=value[,parameter=value]})](https://wiki.mudlet.org/w/Area_51#)
```lua
-- Download from the Internet
loadMusicFile({
    name = "167124__patricia-mcmillen__rugby-club-in-spain.mp3"
    , url = "https://raw.githubusercontent.com/StickMUD/StickMUDSounds/master/sounds/"
})
-- OR download from the profile
loadMusicFile({name = getMudletHomeDir().. "/167124__patricia-mcmillen__rugby-club-in-spain.mp3"})
-- OR download from the local file system
loadMusicFile({name = "C:/Users/Tamarindo/Documents/167124__patricia-mcmillen__rugby-club-in-spain.mp3"})
```

mpconley committed:
Add load/play/stopMedia to Lua API

mpconley committed:
Add load/play/stopMedia to Lua API

mpconley committed:
Add source/load/play/stopMedia to Lua API

mpconley committed:
Add source/load/play/stopMedia to Lua API

mpconley commented:
OK, documented and ready for review.  Using the same platform for media handling as has been done with MSP, MXP and GMCP to expose similar functionality to the Lua API.

The concepts are very much the same other than here it is possible to copy in a sound from the local file system as well to be managed by this framework. Ultimately, a script author should find a place on the Internet to host their sound files, then use `sourceMedia(url)` to capture where to find sounds. The author could preload sounds with `loadMedia(table)`. Next, the script author could `playMedia(table)` and leverage the similar parameters from MCMP (name, type, key, tag, volume, fadein, fadeout, loops, priority, continue), Then the script author could `stopMedia([table])` using filters, such as with name, type, key, tag or priority - or sending no argument will stop all sound and music.

Kebap commented:
> OK, documented and ready for review

Hey thanks for the quick documentation. 

Please note we now collect documentation at a different place in wiki called Area 51 before release, only move it to proper place after release, to avoid confusion 😉

mpconley commented:
This should close #5799

mpconley commented:
This should close #3279

mpconley commented:
Recommend this to close #3049. More control is now in the author's hands to deal with issues like this through various strategies, such as using "key" or "continue".

mpconley commented:
#2997 should be closed as well. That issue has an OS dependency.

mpconley commented:
This will close #2518.

mpconley commented:
This will close #2259.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5786,6 +5787,291 @@ int TLuaInterpreter::purgeMediaCache(lua_State* L)
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "sourceMedia: bad argument #1 type (message as string expected, got %1!)", luaL_typename(L, 1));
+        return lua_error(L);
+    }
```
I'd use `getVerifiedString()` in place of this and then pass the result to `setMediaUrl()` - less duplication of checking logic then. Same in other places.

mpconley replied:
Used getVerifiedXXX just about everywhere I could.  I am interested to see some of these in action - whether it is at compile time or run time?

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5786,6 +5787,291 @@ int TLuaInterpreter::purgeMediaCache(lua_State* L)
+    }
+
+    if (mediaData.getMediaFileName().isEmpty()) {
+        return warnArgumentValue(L, __func__, "loadMedia: missing required information for \"name\" (add index \"name\" and value of file to play)");
```
`warnArgumentValue` already prints the function name, so we can skip that. Suggest simplifying the suggestion to lower technical bar as wel:

```suggestion
        return warnArgumentValue(L, __func__, R"(missing name (add "name" = "file to play"))");
```

mpconley replied:
Shortened this a good bit to eliminate the quoting.  The `R` syntax fools with my vscode a bit, so I avoided it for now and just kept this very simple.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5786,6 +5787,291 @@ int TLuaInterpreter::purgeMediaCache(lua_State* L)
+            }
+        } else if (key == QLatin1String("continue")) {
+            if (!lua_isboolean(L, -1)) {
+                lua_pushfstring(L, "playMedia: bad argument #1 type (for index %s, table values must be booleans, got value: %s (must be boolean))", luaL_typename(L, -2), luaL_typename(L, -1));
```
```suggestion
                lua_pushfstring(L, "playMedia: bad argument #1 type (for index %s, table values must be booleans, got value: %s)", luaL_typename(L, -2), luaL_typename(L, -1));
```

mpconley replied:
Took all of these out I believe.  Many comments like this one are as a result of mirroring things done elsewhere in Mudlet and hoping prior reviews caught them :) Better now!

vadi2 commented on src/TMedia.cpp:
```diff
@@ -57,16 +62,14 @@ void TMedia::playMedia(TMediaData& mediaData)
     mediaData.setMediaFileName(mediaData.getMediaFileName().replace(QLatin1Char('\\'), QLatin1Char('/')));
 
-    if (!TMedia::isFileRelative(mediaData)) {
+    if (mediaData.getMediaProtocol() == TMediaData::MediaProtocolMSP && mediaData.getMediaFileName() == "Off") {
```
```suggestion
    if (mediaData.getMediaProtocol() == TMediaData::MediaProtocolMSP && mediaData.getMediaFileName() == qsl("Off")) {
```

vadi2 commented on src/TMedia.cpp:
```diff
@@ -232,6 +273,25 @@ void TMedia::stopAllMediaPlayers()
+    QDir mediaDir(mediaPath);
+
+    if (!mediaDir.mkpath(mediaPath)) {
+        qWarning() << qsl("TMedia::playMedia() WARNING - Attempt made to create a directory failed: %1").arg(mudlet::getMudletPath(mudlet::profileMediaPath, mpHost->getName()));
```
```suggestion
        qWarning() << qsl("TMedia::playMedia() WARNING - attempt made to create a directory failed: %1").arg(mudlet::getMudletPath(mudlet::profileMediaPath, mpHost->getName()));
```

mpconley replied:
Changed case everywhere

vadi2 commented on src/TMedia.cpp:
```diff
@@ -232,6 +273,25 @@ void TMedia::stopAllMediaPlayers()
+        QFile mediaFile(mediaFilePath);
+
+        if (!mediaFile.exists() && !QFile::copy(mediaData.getMediaFileName(), mediaFilePath)) {
+            qWarning() << qsl("TMedia::playMedia() WARNING - Attempt made to copy file %1 to a directory %2 failed.").arg(mediaData.getMediaFileName(), mediaFilePath);
```
```suggestion
            qWarning() << qsl("TMedia::playMedia() WARNING - attempt made to copy file %1 to a directory %2 failed.").arg(mediaData.getMediaFileName(), mediaFilePath);
```

vadi2 commented on src/TMedia.cpp:
```diff
@@ -275,7 +337,10 @@ bool TMedia::isFileRelative(TMediaData& mediaData)
-        qWarning() << qsl("TMedia::isFileRelative() WARNING - Attempt made to send an absolute path as a media file name: %1.  Only relative paths are permitted.").arg(mediaData.getMediaFileName());
+        if (mediaData.getMediaProtocol() == TMediaData::MediaProtocolMSP || mediaData.getMediaProtocol() == TMediaData::MediaProtocolGMCP) {
+            qWarning()
+                    << qsl("TMedia::isFileRelative() WARNING - Attempt made to send an absolute path as a media file name: %1.  Only relative paths are permitted.").arg(mediaData.getMediaFileName());
```
```suggestion
                    << qsl("TMedia::isFileRelative() WARNING - attempt made to send an absolute path as a media file name: %1. Only relative paths are permitted.").arg(mediaData.getMediaFileName());
```

vadi2 commented on src/lua-function-list.json:
```diff
@@ -268,6 +268,7 @@
     "killTrigger": "killTrigger(id)",
     "loadJsonMap": "loadJsonMap(pathFileName)",
     "loadMap": "loadMap(file location)",
+    "loadMedia": "loadMedia(settings table)",
```
This file is autogenerated from the wiki, so don't worry about populating it.

Did you mean to add the these to the networking functions category in the wiki? playSoundFile is in miscellaneous category, probably should go there.

mpconley replied:
I moved these to Area 51 and we could move them to the same place as playSoundFile is now to make it all better.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5786,6 +5787,291 @@ int TLuaInterpreter::purgeMediaCache(lua_State* L)
 }
 
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#sourceMedia
+int TLuaInterpreter::sourceMedia(lua_State* L)
```
Can this be used to set a location in a package? I can see some folk wanting to skip the effort of hosting sounds online and just exporting a package with their sounds thrown in.

mpconley replied:
Refactored and most eliminated the sourceMedia concept as it was for the API.  People could put the sounds all with their package.  Definitely highlighting this in the documentation now of this PR.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5786,6 +5787,291 @@ int TLuaInterpreter::purgeMediaCache(lua_State* L)
+        lua_pushfstring(L, "sourceMedia: bad argument #1 type (message as string expected, got %1!)", luaL_typename(L, 1));
+        return lua_error(L);
+    }
+
```
Is there a way to see all the source media locations that have been set? That would aid in debugging.

It would also be good to be able to clean up after yourself on package uninstall, so to remove locations.

mpconley replied:
Refactored. For the API, no longer caching a source URL (which is nicer for MSP or MCMP - not needed for the API), we play or load once from a URL, and get the file.  If someone is crossing sounds for some reason, they could search their packages in Mudlet for file references or use the existing purgeMediaCache() to wipe out the profile's media directory and download fresh files.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5786,6 +5787,291 @@ int TLuaInterpreter::purgeMediaCache(lua_State* L)
+}
+
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#playMedia
+int TLuaInterpreter::playMedia(lua_State* L)
```
With old API you can do:

```lua
playSoundFile("/User/bob/desktop/attack.mp3")
```

With new API you have to do:

```lua
playMedia({name = "/User/bob/desktop/attack.mp3"})
```
We don't have to get rid of the convenience of the old one, do we? Would be nice to keep the single argument version + have the more powerful features as well.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5786,6 +5787,291 @@ int TLuaInterpreter::purgeMediaCache(lua_State* L)
+}
+
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#stopMedia
+int TLuaInterpreter::stopMedia(lua_State* L)
```
How does this interact with the already existing `stopSounds()`? Lets document that both in `stopMedia()` and `stopSounds()` on how they work together (or don't, that's ok).

mpconley committed:
Pre-existing bug fix for MSP

mpconley committed:
Fixed: position 0 volume 0 for fadein

mpconley committed:
Feedback: position 0 volume 1 for fadein

vadi2 commented on src/lua-function-list.json:
```diff
@@ -296,6 +297,7 @@
     "permScript": "permScript(name, parent, lua code)",
     "permSubstringTrigger": "permSubstringTrigger( name, parent, pattern table, lua code )",
     "permTimer": "permTimer(name, parent, seconds, lua code)",
+    "playMedia": "playMedia(settings table)",
```
Would be nice to get documentation on all of the keys that playMedia accepts - if you're not coming from an MCMP background, there's not enough context described right now.

Just as an example I set `type = "music", continue = "true"` expecting my sound to loop, but it didn't.

mpconley replied:
Doing a better job at that in this PR and similar text will be in the wiki soon!

vadi2 commented on src/TMedia.cpp:
```diff
@@ -57,16 +62,14 @@ void TMedia::playMedia(TMediaData& mediaData)
-    if (mediaData.getMediaProtocol() == TMediaData::MediaProtocolMSP && mediaData.getMediaFileName() == "Off") {
-        return;
+    if (!fileRelative && (mediaData.getMediaProtocol() == TMediaData::MediaProtocolMSP || mediaData.getMediaProtocol() == TMediaData::MediaProtocolGMCP)) {
+        return; // MSP and GMCP files will not have absolute paths. Something is wrong.
```
```suggestion
        return; // MSP and MCMP files will not have absolute paths. Something is wrong.
```

?

mpconley replied:
Made this change.

vadi2 reviewed (commented):
This is a great improvement to the sound API, really solidifies it. Thanks so much!

mpconley commented:
> > OK, documented and ready for review
> 
> Hey thanks for the quick documentation.
> 
> Please note we now collect documentation at a different place in wiki called Area 51 before release, only move it to proper place after release, to avoid confusion 😉

Moved docs over to Area 51 and I will cross-reference them with the template. Considering some changes to the API now based on feedback so they need to bake a little more.

mpconley commented:
I will probably finish this up over the weekend - a couple of simplification ideas have come to mind from the comments.  Hope to make the release.

mpconley committed:
Significant refactor

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5338,27 +5338,1280 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
+    bool boolValue;
+
+    if (numArgs > 9) {
+        return warnArgumentValue(L, __func__, "playMusicFile: requires no more than 9 arguments");
```
nice to leave this hanging

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5338,27 +5338,1280 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
+    int numArgs = lua_gettop(L);
+
+    if (numArgs > 18) {
+        return warnArgumentValue(L, __func__, "playMusicFile: requires no more than 18 arguments");
```
wont come out in the error right

mpconley committed:
Actions suggestions

mpconley committed:
Late build breaking typo fixed

mpconley committed:
Fixed a suggestion

mpconley commented on src/TLuaInterpreter.cpp:
```diff
@@ -5337,27 +5338,1280 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
+            if (arg1 != QLatin1String("name")) {
+                // values as ordered args: name[,key][,tag][,priority])
+                return TLuaInterpreter::stopSoundsAsOrderedArguments(L);
+            } else {
```
Refactored

mpconley committed:
Reducing

SlySven commented on src/TMedia.cpp:
```diff
@@ -687,7 +768,7 @@ TMediaPlayer TMedia::matchMediaPlayer(TMediaData& mediaData, const QString& abso
                 pPlayer = pTestPlayer;
                 pPlayer.setMediaData(mediaData);
-                pPlayer.getMediaPlayer()->setVolume(mediaData.getMediaVolume());
+                pPlayer.getMediaPlayer()->setVolume(mediaData.getMediaFadeIn() != TMediaData::MediaFadeNotSet ? 1 : mediaData.getMediaVolume());
```
In the previous change (added a warning about being `nable to create new QMediaPlayer object.`) there is a comment that says `// It should be impossible to ever reach this.` - but if it is **not** impossible then this line at least, will cause a seg. fault. So, should it handle it or not?

mpconley replied:
Good to check on, thank you! I simulated this and got the warning to the console and continued operations of Mudlet.

<img width="744" alt="Screen Shot 2022-01-09 at 5 01 40 PM" src="https://user-images.githubusercontent.com/3058178/148702841-16cbc9d3-5eed-41ff-9fee-15df7c501873.png">

mpconley committed:
Added start parameter

mpconley commented:
Ready for final reviews, please!

I incorporated the feedback and improved the API.  We have `playSoundFile()` and `stopSounds()` today.  Existing code keeps working, but it will run on the TMedia engine and benefit from new capabilities.

playSoundFile() gets 8 more lovely parameters.  The first parameter is required, the file `name`, while the rest must be listed in order.  One can stop with 1, 2, 3, 4 or more parameters, but they need to be in the syntax order `playSoundFile(name[,volume][,fadein][,fadeout][,start][,loops][,key][,tag][,priority][,url])`.  If one needs just `name` and `url` they must: `playSoundFile("filename.mp3",nil,nil,nil,nil,nil,nil,nil,nil,"https://example[dot]com/media")`.  As remembering the parameter map is tedious, there are two more flavors of the API available.  I very much like both!  Option 2 is an array of parameters pairs.  The first pair is `"name", "filename.extension"`.  There after, one could list other parameter pairs in any order they like, such as `playSoundFile("name", "filename.mp3","url","https://example[at]com/media/", "priority", 75)`, which is a lot simpler to manage.  Option 3 is similar, in that is uses a table format to do something similar, such as `playSoundFile({"name" = "filename.mp3","url" = "https://example[dot]com/media/"})`.  In the table format, it is not necessary to have the `name` key as the first parameter.  Do we need 3 flavors of syntax?  We need Option 1 for backwards compatibility.  We could choose to reduce to either Option 2 or Option 3.  I love the flexibility here and suggest to keep them as-is.  Summary: Wonderful things to simplify and enhance the user experience could be done leveraging the various parameters once one digs in, and the Internet download capabilities open up possibilities for game admins and package authors who share their scripts with many others.

stopSounds() works as before, but gets 4 more parameters to filter down to stop specific sounds played via playSoundFile().  One could filter by `name`, `key`, `tag`, and `priority`. Similar to playSoundFile(), there are 3 corresponding API syntax flavors.  The more filters provided narrows the choices of matching sounds to stop.  When one lists more than 1 parameter, it is like an AND in a database query, not an OR.  If you want to OR, use more than one stopSounds() call.  Also added is a stopAllSounds() API call, which will be simple to remember with not much more program code spared to add it.

Like Mud Sounds Protocol and Mud Client Media Protocol (MCMP), there is also a suite of the API dedicated to music, which is often the ambience playing in the background, changing as ones game character moves from scene to scene.  We have playMusicFile(), stopMusic() and stopAllMusic(), parallel to what we described above for sound.  Music is slightly different, where it does not have the `priority` parameter, but does add the `continue` parameter.  The `continue` parameter could detect if the same music is already playing.  For example, if one ties music to `gmcp.Room.Info.environment` event changes, they may fire an API called to playMusicFile() specific to that area.  If one moves within the same area, the `"continue" = true` enables the music to keep flowing continuously if another room starts the same music, or if it had stopped, it will play it again.  Separating out sound and music has many benefits for the creative scripter.

stopAllMedia() is there for when one needs immediate quiet from all the API's sounds and music.  And we have loadSoundFile() and loadMusicFile() for preloading files from the Internet or local file system.  These are nice to have, however the url capabilities in playSoundFile() and playMusicFile() should be good enough for most as Internet connections are fast these days in most areas.

Per request, the `start` functionality is new to Mudlet's media capabilities, which begins playing a sound or music a number of milliseconds specified into the track.  Additionally, `start` is available for MCMP over GMCP via `Client.Media.Play.start`.

SlySven commented:
I haven't checked this new code but one problem I have encountered (which, fair enough. might be out of scope for this PR) with the existing scheme - how are paths for files stored?

In the past we have used absolute path names {including for the sound files played by a trigger}. However I have encountered a problem when playing the same profile on multiple OSes in that the paths are NOT necessarily the same on different OSes and that difference was enough to cause trouble (IIRC enough to *hang* Mudlet!) when the sound file was not in the expected place. I suspect it would be beneficial if the path was stored relative to the "`~/.config/mudlet `" location if possible.

{:bulb: I have a backwards compatible workaround in mind, involving storing the location **twice** in the profile the first time with an extra attribute `"isRelative"`, set to `"true"` and the file location written in that manner together with a second one without it with the existing absolute form - existing code would read both but forget the first when it reads the second whereas new versions would use the first in preference...}

mpconley commented:
> I haven't checked this new code but one problem I have encountered (which, fair enough. might be out of scope for this PR) with the existing scheme - how are paths for files stored?
> 
> In the past we have used absolute path names {including for the sound files played by a trigger}. However I have encountered a problem when playing the same profile on multiple OSes in that the paths are NOT necessarily the same on different OSes and that difference was enough to cause trouble (IIRC enough to _hang_ Mudlet!) when the sound file was not in the expected place. I suspect it would be beneficial if the path was stored relative to the "`~/.config/mudlet `" location if possible.
> 
> {💡 I have a backwards compatible workaround in mind, involving storing the location **twice** in the profile the first time with an extra attribute `"isRelative"`, set to `"true"` and the file location written in that manner together with a second one without it with the existing absolute form - existing code would read both but forget the first when it reads the second whereas new versions would use the first in preference...}

When files are gathered, they are stored in the `media` directory, which is reserved for the current profile.  The code uses the existing path logic found elsewhere in Mudlet to do this effectively and consistently.  When a request comes in to play the file, Mudlet checks if it is inside of the media directory first, and if not tries to fetch it from the path presented, which could be a URL, local file or something inside of the profile.  If the file could not be downloaded, the code handles that no worse than other download file operations in Mudlet.  Hopefully we see more URL and "inside the profile" references to files, which will be more OS independent.  If someone is moving between /Users, /home and C:\ often, they may want to consider adding some fancy scripting to handle that :)

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5337,27 +5338,1302 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
+    int numArgs = lua_gettop(L);
+    QString stringValue;
+
+    if (numArgs > 2) {
```
We don't limit the max # of arguments in the API since it is unnecessarily restrictive - doesn't add much value, but it does limit you in case some other function returns more values and you'd like to do `loadSoundFile(otherfunction())`.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5337,27 +5338,1302 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
+}
+
+// Private
+int TLuaInterpreter::loadMediaFileAsOrderedArguments(lua_State* L)
```
I'd expect `loadSoundFile("sdsd")` to give me a nil+msg that the file doesn't exist in the case that it's a local file on the computer. If it's on the web, that would take too long to look up for certain.
 
 Haven't coded this one up.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5337,27 +5338,1302 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
+    }
+
+    if (mediaData.getMediaFileName().isEmpty()) {
+        return warnArgumentValue(L, __func__, QLatin1String("missing argument 1 (file to play)"));
```
API mis-use should be raised as an error, so the developer can spot it straight away while building.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5337,27 +5338,1302 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
+}
+
+// Private
+int TLuaInterpreter::loadMediaFileAsKeyValueArguments(lua_State* L)
```
It's a novel idea, but it's a bit unorthodox... in Lua you typically either use ordered arguments or a table with key-values. Naturally we want the Mudlet API to be clean and tidy, so we'd have to expand this new style to the rest of the API so it's consistent - but adding key value arguments to the entire API would be a lot of work for little gain. I'm for just keeping the ordered and table arguments. There's functions in the API that could use table arguments already, that's enough work to do!

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -5337,27 +5338,1302 @@ int TLuaInterpreter::showUnzipProgress(lua_State* L)
 // Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#playSoundFile
 int TLuaInterpreter::playSoundFile(lua_State* L)
 {
-    QString sound = getVerifiedString(L, __func__, 1, "fileName");
-    if (QDir::homePath().contains('\\')) {
-        sound.replace('/', R"(\)");
-    } else {
-        sound.replace('\\', "/");
```
Do Windows-style paths still work on macOS and Linux with this removal?

A person on Windows could make a package, use Windows-style paths, give it to a friend on macOS and we'd still want it all to work.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -13917,8 +15162,17 @@ void TLuaInterpreter::initLuaGlobals()
+    lua_register(pGlobalLua, "playMusicFile", TLuaInterpreter::playMusicFile);
+    lua_register(pGlobalLua, "stopAllMedia", TLuaInterpreter::stopAllMedia);
+    lua_register(pGlobalLua, "stopAllMusic", TLuaInterpreter::stopAllMusic);
+    lua_register(pGlobalLua, "stopAllSounds", TLuaInterpreter::stopAllSounds);
```
`stopSounds()` does this already - maybe we don't need it `stopAllSounds()` as well?

vadi2 committed:
Vadim's review

vadi2 reviewed (commented):
I had a few changes, and you've done so much already so I've created a PR with them - https://github.com/mpconley/Mudlet/pull/1

demonnic commented:
tl;dr options 1 and 3 please, please leave option 2 out.

In regards to the various API flavors:

I understand we need to keep the ordered parameters with all the nils for backwards compatibility and all that, but it is pretty horrendous to use. Of the other two options, I feel taking a map of options is the best route, it's already done in Geyser so won't be totally unfamiliar, and follows a fairly typical Lua paradigm. It's something I'd like to see added to the other complex functions with unwieldy signatures (looking at you, setTextFormat and tempComplexRegexTrigger) and will be fairly easy to apply consistently across the API. Also, as I recall a table of options is the standard used by sol2 (a Lua binding lib I know we at least briefly considered moving to) and so should be familiar to people.

While the remaining option is intriguing, I would like to humbly request we not include it, as from a support standpoint "shotgun your parameters at this function in almost any order" will be a nightmare, and its inclusion will lead to more confusion than anything else when reading the documentation. We still get folks confused about codestrings vs function references for tempTimer etc despite the latter typically being the preferred practice for years since it was added, just because we still list both for historical and backward compatibility purposes.

I also agree with Vadi that Option 2 would be more difficult to roll out in a consistent manner across the full API, and we should strive to maintain some sort of internal consistency.

mpconley committed:
Merge pull request #1 from vadi2/sound-review

vadim's feedback for sound API

mpconley committed:
homePath() and less stops

vadi2 reviewed (approved):
Nice! Thank you for this massive contribution.

mpconley committed:
Cleanup .h and .json for stopAllXXX()

vadi2 commented:
Congrats on this huge effort @mpconley!
